mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraBlobsDAO.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.james.util.OptionalConverter;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Bytes;
+
+
+public class CassandraBlobsDAO {
+
+    public static final int CHUNK_SIZE = 1024 * 100;
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement insert;
+    private final PreparedStatement insertPart;
+    private final PreparedStatement select;
+    private final PreparedStatement selectPart;
+    private final DataChunker dataChunker;
+
+    @Inject
+    public CassandraBlobsDAO(Session session) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.dataChunker = new DataChunker();
+        this.insert = prepareInsert(session);
+        this.select = prepareSelect(session);
+
+        this.insertPart = prepareInsertPart(session);
+        this.selectPart = prepareSelectPart(session);
+    }
+
+    private PreparedStatement prepareSelect(Session session) {
+        return session.prepare(select()
+            .from(Blobs.TABLE_NAME)
+            .where(eq(Blobs.ID, bindMarker(Blobs.ID))));
+    }
+
+    private PreparedStatement prepareSelectPart(Session session) {
+        return session.prepare(select()
+            .from(BlobParts.TABLE_NAME)
+            .where(eq(BlobParts.ID, bindMarker(BlobParts.ID))));
+    }
+
+    private PreparedStatement prepareInsert(Session session) {
+        return session.prepare(insertInto(Blobs.TABLE_NAME)
+            .value(Blobs.ID, bindMarker(Blobs.ID))
+            .value(Blobs.POSITION, bindMarker(Blobs.POSITION))
+            .value(Blobs.PART, bindMarker(Blobs.PART)));
+    }
+
+    private PreparedStatement prepareInsertPart(Session session) {
+        return session.prepare(insertInto(BlobParts.TABLE_NAME)
+            .value(BlobParts.ID, bindMarker(BlobParts.ID))
+            .value(BlobParts.DATA, bindMarker(BlobParts.DATA)));
+    }
+
+    public CompletableFuture<Optional<BlobId>> save(byte[] data) {
+        if (data == null) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+        BlobId blobId = BlobId.forPayload(data);
+        return saveBlobParts(data, blobId)
+            .thenCompose(partIds -> saveBlobPartsReferences(blobId, partIds))
+            .thenApply(any -> Optional.of(blobId));
+    }
+
+    private CompletableFuture<Stream<Pair<Integer, PartId>>> saveBlobParts(byte[] data, BlobId blobId) {
+        return FluentFutureStream.of(
+            dataChunker.chunk(data, CHUNK_SIZE)
+                .map(pair -> writePart(pair.getRight(), blobId, pair.getKey())
+                    .thenApply(partId -> Pair.of(pair.getKey(), partId))))
```

Comment:
I wonder if that won't be more readable if you put the `.thenApply(partId -> Pair.of(pair.getKey(), partId))))` out of `FluentFutureStream.of()` and do a `.map(partId -> Pair.of(pair.getKey(), partId))))` on the FluentFutureStream.